### PR TITLE
Fix flake8 issues on v19, v20 and v21.0.rst

### LINKED ISSUE #24219: DOC: Fix flake8 issues in doc/source/whatsnew/v0.19.0.rst
xref: #24177

We didn't start validating the format of PEP8 and other code standards in the documentation examples until recently. We still have some files with errors, that we need to skip, and that we should fix, so we can start validating them.

**Two steps are required to replicate the issue**

1. Edit `setup.cfg` in the pandas home, and in the flake8-rst section, remove from the `exclude` list the file `doc/source/whatsnew/v0.24.0.rst`
2. Remove the import block ( `from pandas import *`) from the lines (8-11) of the file `v0.24.0.rst`. 

After that, running the next command will report the errors in the file (note that syntax error usually prevent to validate other errors, and the list of errors to fix can become much longer when the syntax error is fixed (please make sure that you are using `flake8-rst` version 0.7.0 or higher):

```
$ flake8-rst doc/source/whatsnew/v0.19.0.rst 
ource/whatsnew/v0.19.0.rst:141:83: E501 line too long (81 > 79 characters)
source/whatsnew/v0.19.0.rst:162:10: F821 undefined name 'DataFrame'
source/whatsnew/v0.19.0.rst:163:25: E251 unexpected spaces around keyword / parameter equals
source/whatsnew/v0.19.0.rst:163:27: E251 unexpected spaces around keyword / parameter equals
source/whatsnew/v0.19.0.rst:317:5: F821 undefined name 'Timestamp'
source/whatsnew/v0.19.0.rst:325:5: F821 undefined name 'Timestamp'
source/whatsnew/v0.19.0.rst:365:40: E127 continuation line over-indented for visual indent
source/whatsnew/v0.19.0.rst:375:19: W605 invalid escape sequence '\d'
source/whatsnew/v0.19.0.rst:451:13: E201 whitespace after '['
source/whatsnew/v0.19.0.rst:451:60: E202 whitespace before ']'
source/whatsnew/v0.19.0.rst:479:33: E126 continuation line over-indented for hanging indent
source/whatsnew/v0.19.0.rst:479:35: E231 missing whitespace after ','
source/whatsnew/v0.19.0.rst:479:37: E231 missing whitespace after ','
source/whatsnew/v0.19.0.rst:479:39: E231 missing whitespace after ','
source/whatsnew/v0.19.0.rst:479:41: E231 missing whitespace after ','
source/whatsnew/v0.19.0.rst:481:29: E126 continuation line over-indented for hanging indent
source/whatsnew/v0.19.0.rst:481:39: E231 missing whitespace after ','
source/whatsnew/v0.19.0.rst:545:20: E231 missing whitespace after ','
source/whatsnew/v0.19.0.rst:545:22: E231 missing whitespace after ','
source/whatsnew/v0.19.0.rst:1004:92: E501 line too long (95 > 79 characters)
source/whatsnew/v0.19.0.rst:1012:84: E501 line too long (95 > 79 characters)
source/whatsnew/v0.19.0.rst:1061:91: E501 line too long (86 > 79 characters)
source/whatsnew/v0.19.0.rst:1072:83: E501 line too long (86 > 79 characters)
```

Once all the errors are addressed, please open a pull request with the fixes in the file, and removing the file from setup.cfg. If you need to do something that feels wrong to fix an error, please ask in a comment to this issue. Please avoid other unrelated changes, which can be addressed in a separate pull request.
cc: @datapythonista

### LINKED ISSUE #24220: DOC: Fix flake8 issues in doc/source/whatsnew/v0.20.0.rst
xref: #24177

We didn't start validating the format of PEP8 and other code standards in the documentation examples until recently. We still have some files with errors, that we need to skip, and that we should fix, so we can start validating them.

**Two steps are required to replicate the issue**

1. Edit `setup.cfg` in the pandas home, and in the flake8-rst section, remove from the `exclude` list the file `doc/source/whatsnew/v0.20.0.rst`
2. Remove the import block ( `from pandas import *`) from the lines (8-11) of the file `v0.20.0.rst`. 

After that, running the next command will report the errors in the file (note that syntax error usually prevent to validate other errors, and the list of errors to fix can become much longer when the syntax error is fixed (please make sure that you are using `flake8-rst` version 0.7.0 or higher):

```
$ flake8-rst doc/source/whatsnew/v0.20.0.rst 
source/whatsnew/v0.20.0.rst:69:21: E128 continuation line under-indented for visual indent
source/whatsnew/v0.20.0.rst:93:15: E203 whitespace before ':'
source/whatsnew/v0.20.0.rst:93:37: E203 whitespace before ':'
source/whatsnew/v0.20.0.rst:134:42: E231 missing whitespace after ':'
source/whatsnew/v0.20.0.rst:134:57: E231 missing whitespace after ':'
source/whatsnew/v0.20.0.rst:195:12: E251 unexpected spaces around keyword / parameter equals
source/whatsnew/v0.20.0.rst:195:14: E251 unexpected spaces around keyword / parameter equals
source/whatsnew/v0.20.0.rst:196:14: E251 unexpected spaces around keyword / parameter equals
source/whatsnew/v0.20.0.rst:196:16: E251 unexpected spaces around keyword / parameter equals
source/whatsnew/v0.20.0.rst:197:12: E251 unexpected spaces around keyword / parameter equals
source/whatsnew/v0.20.0.rst:197:14: E251 unexpected spaces around keyword / parameter equals
source/whatsnew/v0.20.0.rst:318:8: E124 closing bracket does not match visual indentation
source/whatsnew/v0.20.0.rst:448:21: E124 closing bracket does not match visual indentation
source/whatsnew/v0.20.0.rst:549:32: E231 missing whitespace after ','
source/whatsnew/v0.20.0.rst:549:34: E231 missing whitespace after ','
source/whatsnew/v0.20.0.rst:620:83: E501 line too long (98 > 79 characters)
source/whatsnew/v0.20.0.rst:711:7: E999 SyntaxError: invalid syntax
source/whatsnew/v0.20.0.rst:877:84: E501 line too long (102 > 79 characters)
```

Once all the errors are addressed, please open a pull request with the fixes in the file, and removing the file from setup.cfg. If you need to do something that feels wrong to fix an error, please ask in a comment to this issue. Please avoid other unrelated changes, which can be addressed in a separate pull request.
cc: @datapythonista

### LINKED ISSUE #24222: DOC: Fix flake8 issues in doc/source/whatsnew/v0.21.0.rst
We didn't start validating the format of PEP8 and other code standards in the documentation examples until recently. We still have some files with errors, that we need to skip, and that we should fix, so we can start validating them.

**Two steps are required to replicate the issue**

1. Edit `setup.cfg` in the pandas home, and in the flake8-rst section, remove from the `exclude` list the file `doc/source/whatsnew/v0.21.0.rst`
2. Remove the import block ( `from pandas import *`) from the lines (8-11) of the file `v0.21.0.rst`. 

After that, running the next command will report the errors in the file (note that syntax error usually prevent to validate other errors, and the list of errors to fix can become much longer when the syntax error is fixed (please make sure that you are using `flake8-rst` version 0.7.0 or higher):

```
$ flake8-rst doc/source/whatsnew/v0.11.0.rst 
source/whatsnew/v0.21.0.rst:103:10: E999 SyntaxError: invalid syntax
source/whatsnew/v0.21.0.rst:124:45: E231 missing whitespace after ','
source/whatsnew/v0.21.0.rst:242:83: E501 line too long (91 > 79 characters)
source/whatsnew/v0.21.0.rst:243:54: E226 missing whitespace around arithmetic operator
source/whatsnew/v0.21.0.rst:243:57: E226 missing whitespace around arithmetic operator
source/whatsnew/v0.21.0.rst:252:42: E226 missing whitespace around arithmetic operator
source/whatsnew/v0.21.0.rst:529:12: E231 missing whitespace after ':'
source/whatsnew/v0.21.0.rst:529:21: E231 missing whitespace after ':'
source/whatsnew/v0.21.0.rst:587:21: E231 missing whitespace after ','
source/whatsnew/v0.21.0.rst:587:23: E231 missing whitespace after ','
```

Once all the errors are addressed, please open a pull request with the fixes in the file, and removing the file from setup.cfg. If you need to do something that feels wrong to fix an error, please ask in a comment to this issue. Please avoid other unrelated changes, which can be addressed in a separate pull request.
cc: @datapythonista

### PR DESCRIPTION
**Pull request by thoo**
closes #24219
closes #24220
closes #24222
- [x] run `python make.py --single whatsnew/v0.19.0`, `python make.py --single whatsnew/v0.20.0` etc.
Related to: #24177

### CODE REVIEW

File: doc/source/whatsnew/v0.19.0.rst [hunk 1]
```diff
@@ -319,15 +313,15 @@ These provide date offsets anchored (by default) to the 15th and end of month, a
 
 .. ipython:: python
 
-    Timestamp('2016-01-01') + SemiMonthEnd()
+    pd.Timestamp('2016-01-01') + SemiMonthEnd()
```
Comment (FHaase) [inline doc/source/whatsnew/v0.19.0.rst]: Is it possible to use `pd.offsets.SemiMonthEnd()`?
I find it more descriptive as the reader doesn't have to scroll around to find where `SemiMonthEnd()` came from.

File: doc/source/whatsnew/v0.20.0.rst [hunk 2]
```diff
@@ -684,65 +683,65 @@ data-types would yield different return types. These are now made consistent. (:
...
 
   .. code-block:: ipython
 
-     # Series
-     In [5]: pd.Series([pd.Timestamp('20160101', tz='US/Eastern'),
-                        pd.Timestamp('20160101', tz='US/Eastern')]).unique()
-     Out[5]: array([Timestamp('2016-01-01 00:00:00-0500', tz='US/Eastern')], dtype=object)
+   # Series
+   In [5]: pd.Series([pd.Timestamp('20160101', tz='US/Eastern'),
+                     pd.Timestamp('20160101', tz='US/Eastern')]).unique()
```
Comment (FHaase) [inline doc/source/whatsnew/v0.20.0.rst]: `...:` are missing

File: doc/source/whatsnew/v0.20.0.rst [hunk 3]
```diff
@@ -1086,50 +1085,50 @@ joins, :meth:`DataFrame.join` and :func:`merge`, and the ``.align`` method.
...
-     right
+   left = pd.Index([2, 1, 0])
+   left
+   right = pd.Index([1, 2, 3])
+   right
 
-  Previous Behavior:
+Previous Behavior:
 
   .. code-block:: ipython
```
Comment (FHaase) [inline doc/source/whatsnew/v0.20.0.rst]: Why the indentation? I think the directive needs to start unindented.

File: doc/source/whatsnew/v0.20.0.rst [hunk 4]
```diff
@@ -1141,14 +1140,14 @@ is fixed that allowed this to return a ``Series`` under certain circumstance. (:
...
-                   'col3': [1, 3, 9]})
+   df = pd.DataFrame({'col1': [3, 4, 5],
+                      'col2': ['C', 'D', 'E'],
+                      'col3': [1, 3, 9]})
    df
 
 Previous Behavior:
 
-.. code-block:: ipython
+  .. code-block:: ipython
```
Comment (FHaase) [inline doc/source/whatsnew/v0.20.0.rst]: Same

File: doc/source/whatsnew/v0.21.0.rst [hunk 5]
```diff
@@ -244,8 +244,9 @@ First we set the data:
    import numpy as np
    n = 1000
    df = pd.DataFrame({'Store': np.random.choice(['Store_1', 'Store_2'], n),
-                      'Product': np.random.choice(['Product_1', 'Product_2', 'Product_3'], n),
-                      'Revenue': (np.random.random(n)*50+10).round(2),
+                      'Product': np.random.choice(['Product_1',
+                                                   'Product_2', 'Product_3'], n),
```
Comment (FHaase) [inline doc/source/whatsnew/v0.21.0.rst]: Can each `Product_` be on a seperate line?

File: doc/source/whatsnew/v0.20.0.rst [hunk 6]
```diff
@@ -879,39 +878,39 @@ This is *unchanged* from prior versions, but shown for illustration purposes:
...
-                                [[0, 0, 1, 1], [0, 1, 0, 1]]))
+   df = pd.DataFrame({'value': [1, 2, 3, 4]},
+                     index=pd.MultiIndex([['a', 'b'], ['bb', 'aa']],
+                                         [[0, 0, 1, 1], [0, 1, 0, 1]]))
    df
 
 Previous Behavior:
 
-.. code-block:: python
+  .. code-block:: python
```
Comment (FHaase) [inline doc/source/whatsnew/v0.20.0.rst]: Why this indentation?

**Review by FHaase — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.20.0.rst (hunk 2 above)
Comment (thoo) [reply doc/source/whatsnew/v0.20.0.rst]: Fix.

File: doc/source/whatsnew/v0.20.0.rst (hunk 3 above)
Comment (thoo) [reply doc/source/whatsnew/v0.20.0.rst]: I should not remove the `  Previous Behavior:` . There is an indentation due to a bullet point. But I had some other blocks I shouldn't index. So remove indents in other blocks. 👍

File: doc/source/whatsnew/v0.19.0.rst (hunk 1 above)
Comment (thoo) [reply doc/source/whatsnew/v0.19.0.rst]: Do you mean `pd.tseries.offsets.SemiMonthEnd()`?   I think `SemiMonthEnd()` and `SemiMonthBegin()` are used right after importing from `pandas.tseries.offsets` and that was the only time they were used.  I could be wrong. Let me know if it is the case.

File: doc/source/whatsnew/v0.20.0.rst (hunk 3 above)
Comment (FHaase) [reply doc/source/whatsnew/v0.20.0.rst]: I meant for `  .. code-block:: ipython`. If I'm not mistaken the indentation would create a `\<blockquote>`. If that's wanted, the following code lines have to be indented with at least 3 more spaces than the `.. code-block::`

```
   .. code-block:: python

      code-lines
```

File: doc/source/whatsnew/v0.19.0.rst (hunk 1 above)
Comment (FHaase) [reply doc/source/whatsnew/v0.19.0.rst]: `pd.tseries.offsets` is imported as `pd.offsets`. They are basically the same. But maybe it's better to leave it like `pd.tseries.offsets.SemiMonthEnd()` as I don't know whether it's been the case in `v0.19.0`

File: doc/source/whatsnew/v0.20.0.rst (hunk 3 above)
Comment (thoo) [reply doc/source/whatsnew/v0.20.0.rst]: Done

File: doc/source/whatsnew/v0.20.0.rst (hunk 4 above)
Comment (thoo) [reply doc/source/whatsnew/v0.20.0.rst]: Done

File: doc/source/whatsnew/v0.19.0.rst (hunk 1 above)
Comment (thoo) [reply doc/source/whatsnew/v0.19.0.rst]: The [document](https://pandas.pydata.org/pandas-docs/version/0.19/timeseries.html) for v0.19 also uses a single class name like `BDay` or `DataOffset`. So I will keep it as it is now for consistency.

File: doc/source/whatsnew/v0.20.0.rst [hunk 7]
```diff
@@ -194,13 +188,12 @@ Previously, only ``gzip`` compression was supported. By default, compression of
...
 
-.. code-block:: python
+.. ipython:: python
 
-   url = 'https://github.com/{repo}/raw/{branch}/{path}'.format(
-       repo = 'pandas-dev/pandas',
-       branch = 'master',
-       path = 'pandas/tests/io/parser/data/salaries.csv.bz2',
-   )
+   url = (r'https://github.com/{repo}/raw/{branch}/{path}'
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.20.0.rst]: Why a raw string here?

Comment (thoo) [reply doc/source/whatsnew/v0.20.0.rst]: Fixed now.

File: doc/source/whatsnew/v0.19.0.rst [hunk 8]
```diff
@@ -82,52 +76,51 @@ This also illustrates using the ``by`` parameter to group data before merging.
...
-                               '20160525 13:30:00.075']),
-       'ticker': ['GOOG', 'MSFT', 'MSFT',
-                  'MSFT', 'GOOG', 'AAPL', 'GOOG',
-                  'MSFT'],
-       'bid': [720.50, 51.95, 51.97, 51.99,
-               720.50, 97.99, 720.50, 52.01],
-       'ask': [720.93, 51.96, 51.98, 52.00,
-               720.93, 98.01, 720.88, 52.03]},
-       columns=['time', 'ticker', 'bid', 'ask'])
+    trades = pd.DataFrame({
```
Comment (jreback) [inline doc/source/whatsnew/v0.19.0.rst]: is this the new standard to align not on the ipython (the i), but 1 space in? this seems odd

Comment (jreback) [reply doc/source/whatsnew/v0.19.0.rst]: @datapythonista

Comment (jreback) [reply doc/source/whatsnew/v0.19.0.rst]: this seems inconsistent even in this file

File: doc/source/whatsnew/v0.19.0.rst [hunk 9]
```diff
@@ -478,15 +472,15 @@ Other enhancements
...
-     df = pd.DataFrame({'date': pd.date_range('2015-01-01', freq='W', periods=5),
-                        'a': np.arange(5)},
-                       index=pd.MultiIndex.from_arrays([
-                                [1,2,3,4,5],
-                                pd.date_range('2015-01-01', freq='W', periods=5)],
-                            names=['v','d']))
-     df
-     df.resample('M', on='date').sum()
-     df.resample('M', level='d').sum()
+    df = pd.DataFrame({'date': pd.date_range('2015-01-01', freq='W', periods=5),
```
Comment (jreback) [inline doc/source/whatsnew/v0.19.0.rst]: here its different

File: doc/source/whatsnew/v0.19.0.rst [hunk 10]
```diff
@@ -164,13 +158,13 @@ Using a non-regular, but still monotonic index, rolling with an integer window d
...
 
 
-   dft = DataFrame({'B': [0, 1, 2, np.nan, 4]},
-                   index = pd.Index([pd.Timestamp('20130101 09:00:00'),
-                                     pd.Timestamp('20130101 09:00:02'),
-                                     pd.Timestamp('20130101 09:00:03'),
-                                     pd.Timestamp('20130101 09:00:05'),
-                                     pd.Timestamp('20130101 09:00:06')],
-                                    name='foo'))
+   dft = pd.DataFrame({'B': [0, 1, 2, np.nan, 4]},
```
Comment (jreback) [inline doc/source/whatsnew/v0.19.0.rst]: here too

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.19.0.rst (hunk 10 above)
Comment (thoo) [reply doc/source/whatsnew/v0.19.0.rst]: I think I realigned all to overlap with the first letter "i" or "c". Let me know if I miss any block.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @thoo

**Pull request merged**